src/core/ext/client_channel/subchannel.c

```diff
@@ -788,42 +789,39 @@ grpc_call_stack *grpc_subchannel_call_get_call_stack(
   return SUBCHANNEL_CALL_TO_CALL_STACK(subchannel_call);
 }
 
-static void grpc_uri_to_sockaddr(const char *uri_str,
-                                 grpc_resolved_address *addr) {
-  grpc_uri *uri = grpc_uri_parse(uri_str, 0 /* suppress_errors */);
-  GPR_ASSERT(uri != NULL);
-  if (strcmp(uri->scheme, "ipv4") == 0) {
-    GPR_ASSERT(parse_ipv4(uri, addr));
-  } else if (strcmp(uri->scheme, "ipv6") == 0) {
-    GPR_ASSERT(parse_ipv6(uri, addr));
-  } else {
-    GPR_ASSERT(parse_unix(uri, addr));
-  }
-  grpc_uri_destroy(uri);
-}
-
-void grpc_get_subchannel_address_arg(const grpc_channel_args *args,
+void grpc_get_subchannel_address_arg(grpc_exec_ctx *exec_ctx,
+                                     const grpc_channel_args *args,
                                      grpc_resolved_address *addr) {
-  const char *addr_uri_str = grpc_get_subchannel_address_uri_arg(args);
-  memset(addr, 0, sizeof(*addr));
-  if (*addr_uri_str != '\0') {
-    grpc_uri_to_sockaddr(addr_uri_str, addr);
-  }
-}
-
-const char *grpc_get_subchannel_address_uri_arg(const grpc_channel_args *args) {
   const grpc_arg *addr_arg =
       grpc_channel_args_find(args, GRPC_ARG_SUBCHANNEL_ADDRESS);
   GPR_ASSERT(addr_arg != NULL);  // Should have been set by LB policy.
   GPR_ASSERT(addr_arg->type == GRPC_ARG_STRING);
-  return addr_arg->value.string;
+  memset(addr, 0, sizeof(*addr));
+  if (*addr_arg->value.string != '\0') {
+    grpc_slice addr_slice =
+        grpc_base64_decode(exec_ctx, addr_arg->value.string, 0 /* url_safe */);
+    GPR_ASSERT(GRPC_SLICE_LENGTH(addr_slice) <= GRPC_MAX_SOCKADDR_SIZE);
+    memcpy(addr->addr, (const char *)GRPC_SLICE_START_PTR(addr_slice),
+           GRPC_SLICE_LENGTH(addr_slice));
+    addr->len = GRPC_SLICE_LENGTH(addr_slice);
+    grpc_slice_unref(addr_slice);
+  }
+}
+
+char *grpc_get_subchannel_address_uri_arg(grpc_exec_ctx *exec_ctx,
+                                          const grpc_channel_args *args) {
+  grpc_resolved_address addr;
+  grpc_get_subchannel_address_arg(exec_ctx, args, &addr);
+  return addr.len > 0 ? grpc_sockaddr_to_uri(&addr) : gpr_strdup("");
 }
 
 grpc_arg grpc_create_subchannel_address_arg(const grpc_resolved_address *addr) {
   grpc_arg new_arg;
   new_arg.key = GRPC_ARG_SUBCHANNEL_ADDRESS;
   new_arg.type = GRPC_ARG_STRING;
   new_arg.value.string =
-      addr->len > 0 ? grpc_sockaddr_to_uri(addr) : gpr_strdup("");
+      addr->len > 0 ? grpc_base64_encode(addr->addr, addr->len,
```

Comment:
I don't really see the value of base64-encoding the value here.  The value of encoding it as a human-readable string was to make it easy to debug (originally suggested by @ctiller), but it's just as non-debuggable as base64 as it is to just store the sockaddr value as a pointer with a custom vtable.  The latter seems cleaner.